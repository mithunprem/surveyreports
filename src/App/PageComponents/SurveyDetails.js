import React, { Component, Fragment } from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import Breadcrumb from './BreadCrumb';
import ColourLegendView from './ColourLegendView';
import SurveyMetaData from './SurveyMetaData';
import Themes from './Themes';
import formatSurveyResponses from '../Utils/FormatSurveyResponses';

class SurveyDetails extends Component {

  state = {
    surveyDetails: null
  };

  componentDidMount() {
    const surveyDetails = formatSurveyResponses(this.props.surveyDetails);
    this.setState({ surveyDetails });
  }

  render() {
    const { surveyDetails } = this.state;
    if(!surveyDetails) return null;
    const { name, themes } = surveyDetails;

    return (
      <Fragment>
        <div className="survey-details">
          <Breadcrumb pageName={name} />
          <h4 className="ml-4">{name}</h4>
          <SurveyMetaData data={surveyDetails} />
          <Themes themes={themes} />
          <ColourLegendView />
        </div>
      </Fragment>
    )
  }
}

SurveyDetails.propTypes = {
  surveyDetails: PropTypes.object
}

export default withRouter(SurveyDetails);
