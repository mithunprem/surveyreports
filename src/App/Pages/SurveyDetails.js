import React, { Component, Fragment } from 'react';
import { Breadcrumb, BreadcrumbItem } from 'reactstrap';
import { withRouter } from 'react-router-dom';
import { API_URL } from '../Constants';

class SurveyDetails extends Component {

  state = {
    isSurveyDetailsLoading: false,
    surveyDetails: {}
  };

  componentDidMount() {
    this.setState(
      {
        isSurveyDetailsLoading: true
      },
      async () => {
        const surveyDetails = await this.fetchSurveyDetails();
        this.setState({
          isSurveyResultsLoading: false,
          surveyDetails
        });
      }
    );
  }

  static formatQuestionResponses(response) {
    const { survey_result_detail: surveyDetails } = response;
    let themeAverage = 0;
    let questionAverageRating = 0;

    surveyDetails.themes.forEach(theme => {
      themeAverage = 0;
      theme.questions.forEach(question => {
        questionAverageRating =
          this.calculateAverageRating(question.survey_responses);
        question.averageRating = questionAverageRating;
        themeAverage+= questionAverageRating;
      })
      theme.averageRating = themeAverage / theme.questions.length ;
    });
    return surveyDetails;
  }

  static calculateAverageRating(surveyResponses) {
    let responseAggregate = 0;
    let numberOfResponses = 0;

    if (!surveyResponses || !surveyResponses.length > 0) {
      return null;
    }

    surveyResponses.forEach(({ response_content: response }) => {
      if (response && (response !== "" || response !== " ")) {
        responseAggregate+= Number(response);
        numberOfResponses++;
      }
    });

    return responseAggregate / numberOfResponses;
  }

  fetchSurveyDetails() {
    const url = this.props.location.state.url;
    return fetch(`${API_URL}${url}`)
      .then(response => response.json())
      .then(response => this.constructor.formatQuestionResponses(response))
      .catch(error => console.error('Error:', error));
  }

  render() {
    const { surveyDetails } = this.state;
    if (!surveyDetails) return null;

    const {
      name, participant_count, response_rate,
      submitted_response_count, themes
    } = surveyDetails;

    return (
      <Fragment>
        <Breadcrumb tag="nav" listTag="div">
          <BreadcrumbItem tag="a" href="/">Home</BreadcrumbItem>
          <BreadcrumbItem active tag="span">{name}</BreadcrumbItem>
        </Breadcrumb>
          <h3>Details of {name}</h3>
          <hr className="my-2" />
          <p>
            {participant_count}<br/>
            {response_rate}
          </p>
      </Fragment>
    )

  }
}

export default withRouter(SurveyDetails);
