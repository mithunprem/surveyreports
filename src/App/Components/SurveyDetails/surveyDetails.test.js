import React from 'react';
import { create } from 'react-test-renderer';
import SurveyDetails from '../../Components/SurveyDetails';
import { BrowserRouter } from 'react-router-dom';

describe("SurveyDetails", () => {

  const surveyDetailsSampleData = {
    survey_result_detail: {
      name: "Simple Survey",
      participant_count: 6,
      response_rate: 0.8333333333333334,
      submitted_response_count: 5,
      themes: [
        {
          name: "The Work",
          questions: [
            {
              description: "Test question 1",
              survey_responses: [{response_content: "5"}],
              averageRating: 3.6
            },
            {
              description: "Test question 2",
              survey_responses: [{response_content: "4"}],
              averageRating: 4.6
            }
          ],
          averageRating: 4.866666666666666}
      ]
    }
  };

  const history = {};

  test("should render without throwing an error", () => {
    const div = document.createElement('div');
    div.setAttribute("id", "viewMetaDataButton");
    document.body.appendChild(div);

    const surveyDetails = create(
      <BrowserRouter>
        <SurveyDetails surveyDetails={surveyDetailsSampleData} history={history} />
      </BrowserRouter>).toJSON();
    expect(surveyDetails).toMatchSnapshot();
  });
});
