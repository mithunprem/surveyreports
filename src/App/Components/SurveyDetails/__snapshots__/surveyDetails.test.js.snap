// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyDetails should render without throwing an error 1`] = `
<div
  className="survey-details"
>
  <nav
    aria-label="breadcrumb"
    className=""
  >
    <div
      className="breadcrumb"
    >
      <a
        className="breadcrumb-item"
        href="/"
      >
        Home
      </a>
      <span
        aria-current="page"
        className="active breadcrumb-item"
      >
        Simple Survey
      </span>
    </div>
  </nav>
  <h4
    className="ml-4"
  >
    Simple Survey
  </h4>
  <button
    aria-label={null}
    className="ml-3 btn btn-link btn-sm"
    id="viewMetaDataButton"
    onClick={[Function]}
  >
    View Survey Metadata
  </button>
  <hr />
  <h6
    className="ml-4 mb-2"
  >
     Themes Focussed :
  </h6>
  <div
    className="themes ml-1 mb-3 mr-1 row"
  >
    <div
      className="mb-4 col-12 col-sm-12 col-md-6 col-xl-6"
    >
      <div
        className="theme-details margin-green card card-body"
      >
        <h5
          className="card-header"
        >
          The Work
        </h5>
        <div
          className="card-body"
        >
          <button
            aria-label={null}
            className="btn btn-link btn-sm"
            onClick={[Function]}
            type="button"
          >
            View responses in detail
          </button>
          <div
            className="collapse"
            style={Object {}}
          >
            <table
              className="questions-table table table-striped"
            >
              <thead>
                <tr>
                  <th>
                    Question
                  </th>
                  <th
                    className="average-rating-header"
                  >
                    Average Rating
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td>
                    Test question 1
                  </td>
                  <td
                    className="average-rating-value text-green"
                  >
                    5.00
                  </td>
                </tr>
                <tr>
                  <td>
                    Test question 2
                  </td>
                  <td
                    className="average-rating-value text-green"
                  >
                    4.00
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          className="text-muted card-footer"
        >
          Average rating : 
          4.50
        </div>
      </div>
    </div>
  </div>
  <div
    className="ml-3 mb-4 color-legend"
  >
    <span
      className="small"
    >
      Color legend :
    </span>
    <div
      className="progress"
    >
      <div
        aria-valuemax={100}
        aria-valuemin="0"
        aria-valuenow="33.3"
        className="progress-bar bg-danger"
        role="progressbar"
        style={
          Object {
            "width": "33.3%",
          }
        }
      >
        Rating &lt; 3.5
      </div>
      <div
        aria-valuemax={100}
        aria-valuemin="0"
        aria-valuenow="33.3"
        className="progress-bar bg-warning"
        role="progressbar"
        style={
          Object {
            "width": "33.3%",
          }
        }
      >
        3.5 &lt; Rating &lt; 4
      </div>
      <div
        aria-valuemax={100}
        aria-valuemin="0"
        aria-valuenow="33.3"
        className="progress-bar bg-success"
        role="progressbar"
        style={
          Object {
            "width": "33.3%",
          }
        }
      >
        Rating &gt; 4
      </div>
    </div>
  </div>
</div>
`;
